internal/backend/runtime/omni/controllers/omni/machine_set_node.go

```diff
@@ -28,195 +30,246 @@ import (
+		return res, nil
+	case omni.MachineType, system.ResourceLabelsType[*omni.MachineStatus]():
+		machineSet, err := getMachineSetOfMachine(ctx, r, ptr.ID())
 		if err != nil {
-			return err
+			return nil, err
 		}
 
-		machineMap := make(map[resource.ID]*omni.Machine, allMachines.Len())
+		if machineSet != nil {
+			return machineSet, nil
+		}
 
-		err = allMachines.ForEachErr(func(machine *omni.Machine) error {
-			requestName, ok := machine.Metadata().Labels().Get(omni.LabelMachineRequest)
-			if ok {
-				request, e := safe.ReaderGetByID[*infra.MachineRequest](ctx, r, requestName)
-				if e != nil && !state.IsNotFoundError(e) {
-					return e
-				}
+		// machine is not part of a machine set. Check if machine is allocatable, if so find machineSets that would be interested in the machine.
+		status, err := safe.ReaderGetByID[*machineStatusLabels](ctx, r, ptr.ID())
+		if err != nil {
+			return nil, err
+		}
 
-				if request == nil || request.Metadata().Phase() == resource.PhaseTearingDown {
-					return nil
-				}
-			}
+		selector := resource.LabelQuery{
+			Terms: assignableMachineStatusLabelTerms,
+		}
 
-			machineMap[machine.Metadata().ID()] = machine
+		machineIsPossiblyAssignable := selector.Matches(*status.Metadata().Labels())
 
-			return nil
-		})
-		if err != nil {
-			return err
+		if machineIsPossiblyAssignable {
+			return ctrl.getUpscalableMachinesets(ctx, r)
 		}
 
-		allMachineStatuses, err := safe.ReaderListAll[*machineStatusLabels](ctx, r)
+		return nil, nil
+	case omni.MachineClassType:
+		allMachineSets, err := safe.ReaderListAll[*omni.MachineSet](ctx, r)
 		if err != nil {
-			return err
+			return nil, err
 		}
 
-		machineStatusMap := map[resource.ID]*machineStatusLabels{}
-
-		allMachineStatuses.ForEach(func(ms *machineStatusLabels) {
-			if m, ok := machineMap[ms.Metadata().ID()]; !ok || m.Metadata().Phase() == resource.PhaseTearingDown {
-				ms.Metadata().Labels().Set(labelEvicted, "")
+		var machineSetsWithClass []resource.Pointer
 
-				return
+		allMachineSets.ForEach(func(ms *omni.MachineSet) {
+			allocation := ms.TypedSpec().Value.MachineAllocation
+			if allocation.Source == specs.MachineSetSpec_MachineAllocation_MachineClass && allocation.Name == ptr.ID() {
+				machineSetsWithClass = append(machineSetsWithClass, ms.Metadata())
 			}
-
-			machineStatusMap[ms.Metadata().ID()] = ms
 		})
 
-		visited := map[resource.ID]struct{}{}
+		return machineSetsWithClass, nil
+	case omni.MachineSetNodeType:
+		machineSetNode, err := safe.ReaderGet[*omni.MachineSetNode](ctx, r, ptr)
+		if err != nil {
+			return nil, err
+		}
 
-		err = list.ForEachErr(func(machineSet *omni.MachineSet) error {
-			return ctrl.reconcileMachineSet(ctx, r, machineSet, allMachineStatuses, allMachineSetNodes, machineStatusMap, visited, logger)
-		})
+		machineSetId, ok := machineSetNode.Metadata().Labels().Get(omni.LabelMachineSet)
+		if !ok {
+			// maybe should log a warning. Ask in pr
```

Comment:
Logging would be nice to have. Can also return error here. It will also be logged.